### PR TITLE
New interface for ECCORestoring

### LINKED ISSUE #177: `ECCORestoring` can be used directly as a forcing
Right now, `ECCORestoring` is abstract as the forcing "function", which must then be further wrapped in `Forcing` by `ECCO_restoring_forcing`:

https://github.com/CliMA/ClimaOcean.jl/blob/b3b44b709c70b296adb63cf7b2f4552334dcfd1b/src/DataWrangling/ecco_restoring.jl#L276-L279

There's no barrier to using `ECCORestoring` directly as a forcing though, provided the appropriate functionality within Oceananigans is extended correctly. And this would be much better, not just for syntax but also because the implementation will be less complicated and easier to understand.

### PR DESCRIPTION
**Pull request by simone-silvestri**
should close #177

This PR also includes a `maxiter` field to the ECCONetCDFbackend that controls the maximum number of iterations for the inpainting algorithm

### CODE REVIEW

File: src/DataWrangling/ECCO/ECCO_mask.jl [hunk 1]
```diff
@@ -40,3 +41,48 @@ end
...
+struct LatitudinallyTaperedPolarMask{N, S, Z} 
+    northern_edges :: N
+    southern_edges :: S
+    z_edges :: Z
+end
+
+"""
+    LatitudinallyTaperedPolarMask(; northern_edges = (70,   75),
+                                    southern_edges = (-75, -70),
+                                    z_edges = (-20, 0))
```
Comment (glwagner) [inline src/DataWrangling/ECCO/ECCO_mask.jl:53]: Nice! Is the proposed syntax change intentional? It wasn't mentioned in the description...

File: src/DataWrangling/ECCO/ECCO_restoring.jl [hunk 2]
```diff
@@ -20,43 +21,41 @@ import Oceananigans.OutputReaders: new_backend, update_field_time_series!
 struct ECCONetCDFBackend{N} <: AbstractInMemoryBackend{Int}
     start :: Int
     length :: Int
+    maxiter :: Int
```
Comment (glwagner) [inline src/DataWrangling/ECCO/ECCO_restoring.jl:24]: should we make it more clear this refers to inpainting?

File: src/DataWrangling/ECCO/ECCO_restoring.jl [hunk 3]
```diff
@@ -173,22 +176,22 @@ A struct representing ECCO restoring.
 - `ECCO_grid`: The native ECCO grid to interpolate from.
 - `mask`: A mask (could be a number, an array, a function or a field).
 - `variable_name`: The variable name of the variable that needs restoring.
-- `λ⁻¹`: The reciprocal of the restoring timescale.
+- `rate`: The reciprocal of the restoring timescale.
 """
-struct ECCORestoring{FTS, G, M, V, N} <: Function
+struct ECCORestoring{FTS, G, M, V, N} 
     ECCO_fts      :: FTS
     ECCO_grid     :: G
```
Comment (glwagner) [inline src/DataWrangling/ECCO/ECCO_restoring.jl:183]:
```suggestion
    field_time_series      :: FTS
    grid     :: G
```

File: src/DataWrangling/ECCO/ECCO_restoring.jl [hunk 4]
```diff
@@ -231,52 +234,65 @@ end
...
-function ECCO_restoring_forcing(metadata::ECCOMetadata;
-                                architecture = CPU(), 
-                                time_indices_in_memory = 2, # Not more than this if we want to use GPU!
-                                time_indexing = Cyclical(),
-                                mask = 1,
-                                timescale = 20days,
-                                grid = nothing)
+function ECCORestoring(metadata::ECCOMetadata;
+                       architecture = CPU(), 
+                       time_indices_in_memory = 2, # Not more than this if we want to use GPU!
```
Comment (glwagner) [inline src/DataWrangling/ECCO/ECCO_restoring.jl:281]: should this be `backend`?

File: src/DataWrangling/ECCO/ECCO_mask.jl (hunk 1 above)
Comment (simone-silvestri) [reply src/DataWrangling/ECCO/ECCO_mask.jl:53]: I am actually ok also with `LinearlyTaperedPolarMask`. I just thought `Latitudinally` would express the fact that longitude doesn't play a role with the mask. However, I am happy to switch to `LinearlyTaperedPolarMask`

Comment (glwagner) [reply src/DataWrangling/ECCO/ECCO_mask.jl:53]: yes and also adding the clarification `edges` to every argument. Somehow I think we want to make it so we don't need that clarification, perhaps by making the name of the type better. Notice that every property says "edges", suggesting that there should be a better way

File: test/test_ecco.jl [hunk 5]
```diff
@@ -38,14 +38,67 @@ using Dates
...
+
+    mask = LatitudinallyTaperedPolarMask(northern_edges = (φ₃, φ₄), 
+                                         southern_edges = (φ₁, φ₂), 
+                                         z_edges = (z₁, 0))
+
+    t_restoring = ECCORestoring(:temperature, CPU(); 
+                                dates, 
+                                mask, 
+                                rate = 1 / 1000.0,
+                                maxiter = 1)
```
Comment (glwagner) [inline test/test_ecco.jl:65]: should the max iter be set to the order of the advection scheme? or something like that?

File: src/DataWrangling/ECCO/ECCO_restoring.jl (hunk 4 above)
Comment (simone-silvestri) [reply src/DataWrangling/ECCO/ECCO_restoring.jl:281]: At the moment, the only supported backend is the `ECCOBackendNetCDF`. If we want to be able to preprocess data we can change this to accept a backend and support different backends.

File: src/DataWrangling/ECCO/ECCO_restoring.jl (hunk 2 above)
Comment (simone-silvestri) [reply src/DataWrangling/ECCO/ECCO_restoring.jl:24]: I have changed it to `inpainting_maxiter`. I am not super happy about it though. Do you have a suggestion for a name for this variable?

File: test/test_ecco.jl (hunk 5 above)
Comment (simone-silvestri) [reply test/test_ecco.jl]: I am not sure. I don't think it has to do with the model detail, it has more to do with the discrepancy of the bathymetry we use with the ECCO bathymetry

File: src/DataWrangling/ECCO/ECCO_restoring.jl (hunk 2 above)
Comment (glwagner) [reply src/DataWrangling/ECCO/ECCO_restoring.jl:24]: How about making a new struct, `Inpainting`, which has a property `maxiter` (and which may be generalized to include more settings in the future)

**Comment by simone-silvestri** [comment]
should be mergeable

**Review by glwagner — APPROVED** [review]

File: test/runtests.jl [hunk 6]
```diff
@@ -46,7 +49,7 @@ if test_group == :turbulent_fluxes || test_group == :all
 end
 
 if test_group == :bathymetry || test_group == :all
-    include("test_bathymetry.jl")
+    CUDA.set_runtime_version!(v"12.6"; local_toolkit = true)
```
Comment (glwagner) [inline test/runtests.jl]: you may want to keep the test file though

**Pull request merged**